File: 0139-word-break/0139-word-break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        bool dp[s.size()];
        memset(dp, false, sizeof(dp));
        unordered_set<string> set;
        for(string ss : wordDict) set.insert(ss);
        for(int i=0; i<s.size(); i++) {
            for(int j=i; j<s.size(); j++) {
                string t = s.substr(i, j-i+1);
                if(set.find(t) != set.end()) {
                    if(i-1 >= 0 && dp[i-1]) dp[j] = true;
                    else if(i == 0) dp[j] = true;
                }
            }
        }
        return dp[s.size()-1];
    }
};

File: 0139-word-break/0139-word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        book = set(wordDict)
        #for word in wordDict: book.add(word)
        dp = [False for i in range(len(s))]
        for i in range(len(s)):
            for j in range(i, len(s)):
                if dp[j]: continue
                t = s[i:j+1]
                if t in book:
                    if j-i >= 0 and dp[i-1]: dp[j] = True
                    elif i == 0: dp[j] = True
        return dp[len(s)-1]